core/trino-main/src/main/java/io/trino/operator/MultiChannelGroupByHash.java

```diff
@@ -507,6 +518,24 @@ private boolean canProcessDictionary(Page page)
         return true;
     }
 
+    private boolean canProcessLowCardinalityDictionary(Page page)
+    {
+        // We don't have to rely on 'optimizer.dictionary-aggregations' here since there is little to none chance of regression
+        int positionCount = page.getPositionCount();
+        long cardinality = 1;
+        for (int channel : channels) {
+            if (!(page.getBlock(channel) instanceof DictionaryBlock)) {
+                return false;
+            }
+            cardinality = multiplyExact(cardinality, ((DictionaryBlock) page.getBlock(channel)).getDictionary().getPositionCount());
+            if (cardinality > positionCount * SMALL_DICTIONARIES_MAX_CARDINALITY_RATIO) {
```

Comment:
you should still make sure cardinality is <= https://docs.oracle.com/javase/7/docs/api/java/lang/Short.html#MAX_VALUE. Otherwise it won't fit in `short[] combinationIds = new short[positionCount];`
Add test?